content/http-handler-abi.md

```diff
@@ -767,7 +779,8 @@ method to "POST".
 ;; get_uri writes the URI to memory if it isn't larger than `buf_limit`, e.g.
 ;; "/v1.0/hi?name=panda". The result is its length in bytes.  
 ;;
-;; Note: The URI may include query parameters.
+;; Note: The URI may include query parameters. Despite the path and query being
```

Comment:
This is the important change.

Reply:
BTW I'd also lean towards non-encoded for URI here, it's generally what's used inside middleware like a routing one. Either way, guest SDKs could provide both but the ABI could focus on the commonly used case I guess.

Reply:
I would also like to clarify that even if I suggested we could offer two abi to get headers I really hope we don't because decisions like this could explode the ABI. In other words, a small ABI is a feature and there should be a very good reason for each new function to exist.

For starters, I'll make the impl consistent on raw and hopefully we can re-use this PR to document that and why. 

Here's a pitch on RATIONALE based on @anuraaga's note and some things I was thinking about similarly.

URI is returned raw because that's what's used in routing. An http-handler is an inbound processor, and HTTP properties such as HTTP path and query are often used for routing properties. If we returned the URI decoded, the guest would have higher overhead decoding these in order to route based on them.

Moreover, returning the URI raw from `get_uri` matches that we accept it raw in `set_uri`. Even if a host chooses to coerce an encoded input to `set_uri`, having the requirement of input matching output in the specified case is easier to reason with than otherwise.

Reply:
Tried writing some code using getURI and realized this probably does need to be encoded, and maybe what JC had in mind. e.g., if getting `/disney?name=chip&dale` this would be two separate parameters when unencoded though most likely the encoded version was `/disney?name=chip%26dale`. Having raw values seems to only make sense when splitting the path and query but not when they're together because of the semantic meaning of `&`.

So I guess we can let `get_uri` and `set_uri` in the ABI deal with encoded, and let guest SDKs expose something easier, e.g. in tinygo SDK maybe we can replace `GetURI() string` with `GetURI() URL`. Haven't completely vetted this idea yet

Reply:
ok sounds like progress towards rationale.

I guess if we are going to encode the query, we'd do that with the query encoding which I always forget has slightly different rules than url. For years, I look at okhttp for some cheat sheet examples..

https://github.com/square/okhttp/blob/58ee1ce170c2aea75a113012956260873e808dad/okhttp/src/jvmMain/kotlin/okhttp3/HttpUrl.kt

Reply:
I think recommending types is not a good idea due to how compilers bloat things and also unnecessary overhead. 

Remember in many languages there are alternative HTTP libraries and they use things like buffers or strings to hydrate their URI types as well. Some may want to use a nonstd library and still have to suffer through the bloat using whatever the URI is (especially Java's is bad).

Reply:
case in point is the WIP tinygo thing bloated usage of get_uri by 89KB (1200%)

I think the thinking is better to be lean on types by default as it keeps nervous types like me from having to keep watching to make sure the thing doesn't go sideways.